File: problems/data-structures-n-algorithms/wip/198.c
// 198. House Robber
// https://leetcode.com/problems/house-robber/

#include <stdio.h>
#include <stdlib.h>

int max(int a, int b) {
	if (a > b) {
		return a;
	}
	return b;
}

// int _rob(int *nums, int i) {
// 	if (i < 0) {
// 		return 0;
// 	}

// 	return max(_rob(nums, i - 2) + nums[i], _rob(nums, i - 1));
// }

// int rob(int *nums, int numsSize) {
// 	if (numsSize == 0) {
// 		return 0;
// 	}

// 	return max(_rob(nums, numsSize - 3) + nums[numsSize - 1],
// 						 _rob(nums, numsSize - 1));
// }

int rob(int *nums, int numsSize) {
	int sum1 = 0, sum2 = 0;
	int i;

	for (i = 0; i < numsSize; i++) {
		if (i % 2 == 0) {
			sum1 += nums[i];
		} else {
			sum2 += nums[i];
		}
	}

	return max(sum1, sum2);
}

int main() {
	int *nums, numsSize;
	int i, res;

	scanf("%d", &numsSize);

	nums = (int *)malloc(sizeof(int) * numsSize);

	for (i = 0; i < numsSize; i++) {
		scanf("%d", &nums[i]);
	}

	res = rob(nums, numsSize);

	printf("%d", res);

	return 0;
}

File: problems/data-structures-n-algorithms/medium/198.java
// 198. House Robber
// https://leetcode.com/problems/house-robber/

class S198 {
	public static void main (String args[]) {
		int[] nums = new int[]{1, 2, 3, 1};
		// int[] nums = new int[]{2, 7, 9, 3, 1};
		// int[] nums = new int[]{2, 1, 1, 2};
		int res = rob(nums);
		System.out.println(res);
	}

	public static int rob (int[] nums) {
		int len = nums.length;
		int[] dp = new int[len];
		if (len == 1) {
			return nums[0];
		} else if (len == 2) {
			return Math.max(nums[0], nums[1]);
		}
		dp[0] = nums[0];
		dp[1] = Math.max(nums[0], nums[1]);

		for (int i = 2; i < len; i++) {
			dp[i] = Math.max(dp[i - 2] + nums[i], dp[i - 1]);
		}

		return dp[len - 1];
	}
}